src/main/scala/mesosphere/marathon/core/appinfo/impl/DefaultInfoService.scala

```diff
@@ -29,91 +43,125 @@ private[appinfo] class DefaultInfoService(
+      infos
+    }
 
+  @SuppressWarnings(Array("all")) // async/await
   override def selectAppsInGroup(groupId: PathId, selector: AppSelector,
-    embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] = {
-    log.debug(s"queryAllInGroup $groupId")
-    groupManager
-      .group(groupId)
-      .map(_.map(_.transitiveApps.filter(selector.matches)).getOrElse(Seq.empty))
-      .flatMap(resolveAppInfos(_, embed))
-  }
+    embed: Set[AppInfo.Embed]): Future[Seq[AppInfo]] =
 
-  override def selectGroup(groupId: PathId, groupSelector: GroupSelector,
+    async { // linter:ignore UnnecessaryElseBranch
+      log.debug(s"queryAllInGroup $groupId")
+      val maybeGroup: Option[Group] = await(groupManager.group(groupId))
+      val maybeApps = maybeGroup.map(_.transitiveApps.filter(selector.matches).toVector)
+      maybeApps match {
+        case Some(selectedApps) => await(resolveAppInfos(selectedApps, embed))
+        case None => Seq.empty
+      }
+    }
+
+  override def selectGroup(groupId: PathId, selectors: GroupInfoService.Selectors,
     appEmbed: Set[Embed], groupEmbed: Set[GroupInfo.Embed]): Future[Option[GroupInfo]] = {
     groupManager.group(groupId).flatMap {
-      case Some(group) => queryForGroup(group, groupSelector, appEmbed, groupEmbed)
+      case Some(group) => queryForGroup(group, selectors, appEmbed, groupEmbed)
       case None => Future.successful(None)
     }
   }
 
-  override def selectGroupVersion(groupId: PathId, version: Timestamp, groupSelector: GroupSelector,
+  override def selectGroupVersion(groupId: PathId, version: Timestamp, selectors: GroupInfoService.Selectors,
     groupEmbed: Set[GroupInfo.Embed]): Future[Option[GroupInfo]] = {
     groupManager.group(groupId, version).flatMap {
-      case Some(group) => queryForGroup(group, groupSelector, Set.empty, groupEmbed)
+      case Some(group) => queryForGroup(group, selectors, Set.empty, groupEmbed)
       case None => Future.successful(None)
     }
   }
 
+  @SuppressWarnings(Array("all")) // async/await
   private[this] def queryForGroup(
     group: Group,
-    groupSelector: GroupSelector,
+    selectors: GroupInfoService.Selectors,
     appEmbed: Set[AppInfo.Embed],
-    groupEmbed: Set[GroupInfo.Embed]): Future[Option[GroupInfo]] = {
+    groupEmbed: Set[GroupInfo.Embed]): Future[Option[GroupInfo]] =
 
-    //fetch all transitive app infos with one request
-    val appInfos = {
-      if (groupEmbed(GroupInfo.Embed.Apps))
-        resolveAppInfos(group.transitiveApps.filter(groupSelector.matches), appEmbed)
-      else
-        Future.successful(Seq.empty)
-    }
+    async { // linter:ignore UnnecessaryElseBranch
+      val groupEmbedApps = groupEmbed(GroupInfo.Embed.Apps)
+      val groupEmbedPods = groupEmbed(GroupInfo.Embed.Pods)
+
+      //fetch all transitive app infos and pod statuses with one request
+      val specStatuses: Seq[SpecStatus] = await {
+        resolveSpecInfos(group.transitiveRunSpecs.collect {
```

Comment:
This will resolveSpecInfos for all apps and pods no matter which embed option you have.
The original goal was, to only resolve this data, if the embed option is set.

Since you split the data after resolving anyways, I think separating the 2 cases would be easier.